### PR TITLE
Koa v2.x

### PR DESCRIPTION
**Pull request by jonathanong**
Has been published as `v2.0.0`!

NOTE: keeping this open for visibility since it's hard to see when a repo has multiple working branches

---

Keeping this open for visibility. If anyone would like to help migrating new v1.x commits to v2.x, that would be great!

also, cherry-picking specific commits to master immediately would be really helpful too. for example, merging eslint to master would make merging easier.

merge base: https://github.com/koajs/koa/compare/96c1e0998f1daa9bf1e33b5b0d79209f29351a9d...master

to cherry-pick:
- [x] https://github.com/koajs/koa/commit/ec477107d56d401dcb54057a8648d0247c05abfa
- ~~https://github.com/koajs/koa/commit/bbfa6777a09e1c79045f13e45f76fe4c29109bb8~~
- [x] https://github.com/koajs/koa/commit/0c9c56cc049ab7105d695205e4ab35f362748cf6
- [x] https://github.com/koajs/koa/commit/8cbff160cab0f7f0ef4da47f00cae17f66e0e834
- [x] https://github.com/koajs/koa/commit/a25ab12116cf5b0dd608e009a01c700278a9b3e7
- [x] https://github.com/koajs/koa/commit/d3066715e3ea8c8acb0665df4d8d5e5ba6c6b0d5
- [x] https://github.com/koajs/koa/commit/97c34d19058fc1f5486892389d0457b9ccef8c7b
- [x] https://github.com/koajs/koa/commit/229641127a1f2c6a436d6d633f9d97718b16c298
- [x] https://github.com/koajs/koa/commit/b5c1a60ac3016eab89d7ae98dd33a21134ad4708
- [x] https://github.com/koajs/koa/commit/572b4599f0852e1e5a30a7def245ce88cf5e9339
- [x] https://github.com/koajs/koa/commit/7a461b3a531573118dee54e82edb8e1f07b49398
- [x] https://github.com/koajs/koa/commit/c82646756053395e463cad059ed776611a99c1fd
- [x] https://github.com/koajs/koa/commit/37ea485bc03b3dbca7e92518ad126679e4f76919
- [x] https://github.com/koajs/koa/commit/7bc6e79da345168137b0cc0ab77bacb4a0db15ea

Checked until https://github.com/koajs/koa/commit/7bc6e79da345168137b0cc0ab77bacb4a0db15ea

Other:
- [x] update master docs
- [x] update v2 docs & close relevant doc issues
- [x] update changelog

### CODE REVIEW

**Comment by jonathanong** [comment]
think i'm gonna make a `v2.x-merge` branch later where i'm rebasing. sigh so much work

File: package.json [hunk 1]
```diff
@@ -42,10 +41,18 @@
     "type-is": "^1.5.5",
     "vary": "^1.0.0"
   },
+  "publishConfig": {
```
Comment (jonathanong) [inline package.json]: cool i didn't know about this!

**Comment by jonathanong** [comment]
^ i've cherry-picked some commits from master. help to cherry-pick / merge other relevant ones here would be great :)

**Comment by jonathanong** [comment]
we should decide on https://github.com/koajs/koa/pull/683 and https://github.com/koajs/koa/pull/605 before releasing. please write opinions if you have any there

**Comment by jonathanong** [comment]
went through a first pass of the docs. wanna do another one before i release. would anyone else mind taking a pass at the docs?

**Comment by PlasmaPower** [comment]
I've looked through the docs. I made a PR to fix a minor thing, and the koa vs express document still references generators/co, but it does have WIP at the top. Other than that, looks good.

**Comment by jonathanong** [comment]
:+1: kind of want to remove the koa vs. express doc but it doesn't matter because it doesn't show up on koajs.com

**Comment by fl0w** [comment]
#691 is kind of weird to have right now, references the wiki which in turn has deprecated middleware (e.g. hello world mw and sub-domains). They are also not under koajs gh org.

**Comment by jonathanong** [comment]
@fl0w yeah, i prefer if we updated the wiki vs. removing the link

**Comment by jonathanong** [comment]
k publishing

**Comment by ranisalt** [comment]
So how's the progress on Koa 2? Any predictions on when it will be launched as stable?

**Comment by dominhhai** [comment]
@ranisalt check here https://github.com/koajs/koa/issues/533

**Comment by jonathanong** [comment]
@dead-horse can you double check the changelog for alpha.4 is correct?

**Comment by jonathanong** [comment]
oh i see i'm confused by the changelog now. hahaha.

**Comment by jonathanong** [comment]
@dead-horse can you clarify the versioning scheme?

File: lib/request.js [hunk 2]
```diff
@@ -5,6 +5,7 @@
  * Module dependencies.
  */
 
+const net = require('net');
 const contentType = require('content-type');
```
Comment (fundon) [inline lib/request.js]: About `const isIP = require('net'). isIP` or `const { isIP } = require('net')`

File: Readme.md [hunk 3]
```diff
@@ -5,69 +5,136 @@
...
+  Expressive HTTP middleware framework for node.js to make web applications and APIs more enjoyable to write. Koa's middleware stack flows in a stack-like manner, allowing you to perform actions downstream then filter and manipulate the response upstream.
 
-  Only methods that are common to nearly all HTTP servers are integrated directly into Koa's small ~550 SLOC codebase. This
+  Only methods that are common to nearly all HTTP servers are integrated directly into Koa's small ~570 SLOC codebase. This
   includes things like content negotiation, normalization of node inconsistencies, redirection, and a few others.
 
   Koa is not bundled with any middleware.
 
 ## Installation
+Koa requires __node v4.0.0__ or higher for (partial) ES2015 support.
```
Comment (ilkkao) [inline Readme.md]: Maybe even strongly recommend node 7.6 and have examples in async/await syntax.

**Pull request merged**